tracing-subscriber/src/thread.rs

```diff
@@ -0,0 +1,191 @@
+use std::{
+    cell::{Cell, UnsafeCell},
+    fmt,
+    marker::PhantomData,
+};
+use std::{
+    ops::{Deref, DerefMut},
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+use crossbeam_utils::sync::{ShardedLock, ShardedLockReadGuard};
+
+pub struct Local<T> {
+    inner: ShardedLock<Inner<T>>,
+}
+
+type Inner<T> = Vec<Option<UnsafeCell<T>>>;
+
+pub struct LocalGuard<'a, T> {
+    inner: *mut T,
+    /// Hold the read guard for the lifetime of this guard. We're not actually
+    /// accessing the data behind that guard, but holding the read lock will
+    /// keep another thread from reallocating the vec.
+    _lock: ShardedLockReadGuard<'a, Inner<T>>,
+}
+
+/// Uniquely identifies a thread.
+#[repr(transparent)]
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub(crate) struct Id {
+    id: usize,
+    _not_send: PhantomData<UnsafeCell<()>>,
+}
+
+// === impl Local ===
+
+impl<T> Local<T> {
+    pub fn new() -> Self {
+        let len = Id::current().as_usize();
+        // Preallocate up to the current thread ID, so we don't have to inside
+        // the lock.
+        let mut data = Vec::with_capacity(len);
+        data.resize_with(len, || None);
+        Local {
+            inner: ShardedLock::new(data),
+        }
+    }
+
+    fn get2<'a>(&'a self, i: usize) -> Option<LocalGuard<'a, T>> {
```

Comment:
what is the difference between this one and the on below?

Reply:
this is an internal implementation detail that's used by both `get` and `get_or_else`. maybe it should be renamed.